pkg/downwardmetrics/virtio-serial/server.go

```diff
@@ -0,0 +1,244 @@
+package virtio_serial
+
+import (
+	"bufio"
+	"encoding/xml"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/textproto"
+	"net/url"
+	"path/filepath"
+	"strings"
+	"syscall"
+	"time"
+
+	"kubevirt.io/kubevirt/pkg/downwardmetrics"
+	"kubevirt.io/kubevirt/pkg/util"
+	"kubevirt.io/kubevirt/pkg/virt-handler/isolation"
+
+	"github.com/fsnotify/fsnotify"
+	"kubevirt.io/client-go/log"
+
+	diskutils "kubevirt.io/kubevirt/pkg/ephemeral-disk-utils"
+	metricsScraper "kubevirt.io/kubevirt/pkg/monitoring/domainstats/downwardmetrics"
+)
+
+const standardLauncherSocketPath = "/run/kubevirt/sockets/launcher-sock"
+
+var (
+	invalidRequest = []byte("INVALID REQUEST\n\n")
+	emptyMetrics   = []byte("<metrics><!-- host metrics not available --><!-- VM metrics not available --></metrics>")
+)
+
+func reply(conn net.Conn, response []byte) error {
+	_, err := conn.Write(response)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func replyError(conn net.Conn) {
+	_ = reply(conn, invalidRequest)
+}
+
+func parseRequestLine(line string) (string, error) {
+	method, rawUri, ok := strings.Cut(line, " ")
+	if !ok {
+		return "", errors.New("malformed request")
+	}
+
+	if method != "GET" {
+		return "", fmt.Errorf("invalid method %q", method)
+	}
+
+	requestUri, err := url.ParseRequestURI(rawUri)
+	if err != nil {
+		return "", err
+	}
+
+	return requestUri.Path, nil
+}
+
+func waitForRequest(conn net.Conn) (string, error) {
+	r := textproto.NewReader(bufio.NewReader(conn))
+
+	request, err := r.ReadLine()
+	if err != nil {
+		return "", err
+	}
+	return request, nil
+}
+
+func waitForSocket(socketPath string) error {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return err
+	}
+	defer func(watcher *fsnotify.Watcher) {
+		_ = watcher.Close()
+	}(watcher)
+
+	err = watcher.Add(filepath.Dir(socketPath))
+	if err != nil {
+		return err
+	}
+
+	for {
+		exists, err := diskutils.FileExists(socketPath)
+		if err != nil {
+			return err
+		}
+
+		if exists {
+			break
+		}
+
+		select {
+		case <-watcher.Events:
+		case err, ok := <-watcher.Errors:
+			if !ok {
+				return errors.New("unknown error")
+			}
+			return err
+		}
+	}
+	return nil
+}
+
+func connect(socketPath string, retries uint) (net.Conn, error) {
+	var conn net.Conn
+	var err error
+
+	multiplier := 1
+	for i := uint(0); i < retries+1; i++ {
```

Comment:
Is the explicit uint type necessary?

Reply:
I changed that to just attempts since it's confusing

Reply:
without the uint(0) I can't compile it locally, but it seems to work in the CI. I'll leave that one since I think it's harmless